### ISSUE TITLE
API: Request compression error response

### ISSUE THREAD

**Issue by Endimion75** [Scan-o-Matic/scanomatic#134]
When I make a request to the API for a colony compression:

`/api/data/ccc<ccc_identifier>/image/<image_identifier>/plate/<plate_number>/compress/colony/<x>/<y>`

If I send colony data that is "wrong" I.E. blob values, neither values, but no background values.  Instead of getting a json response like so:

{
  "is_endpoint": false, 
  "reason": "Background must be consisting of at least 3 pixels", 
  "success": false
}

I get a BAD REQUEST error (status: 400) with a response object with the values I was expecting:

{
  "is_endpoint": false, 
  "reason": "Background must be consisting of at least 3 pixels", 
  "success": false
}

Should I be getting directly a success response, but a success = false json object, rather than an error response?

**Comment by skymandr** [comment]
This may also have been fixed in the new pull request. Could you check if the problem persists once it is up?

**Comment by Endimion75** [comment]
will do

**Comment by skymandr** [comment]
There, reset the master branch and made a proper pull request. Please update your code accordingly (currently no changes, but I will commit updates to any other issues to the new branch instead - I have also assigned you as reviewer(s) so please have a look at the change to see that I haven't done anything atrocious.)

**Comment by skymandr** [comment]
This behaviour is as it should be. We will successively make sure that all failed calls return something else than 200 with a description of the error. What may change is that we will likely move away from 400 across the board to more informative 40X status codes. Closing this issue.

**Comment by Endimion75** [comment]
I have problems with this.
I have been constructing the CCC UI with the expectation that the API will behave in certain way.  Suddenly changing the way the API responds without consideration for backwards compatibility does not sparkle with me.

**Comment by Endimion75** [comment]
I understand to return an error if the data sent was misconstrued, incorrect or missing.  But the data is correct, the API could not accommodate properly the grid, but still returns values and a fails state.  

In the past this cases were treated as successful calls with an success = false state.  And ALL of the API calls I make expect that type of behaviour. 

Now are you telling me that all that is to be overhauled?

**Comment by Endimion75** [comment]
Also, if unsuccessful requests are going to be treated as error responses, what is the point of having a success = true/false field in the response?

**Comment by skymandr** [comment]
Ok, so this was definitely a change that should have been a separate pull request, not included in #135... Lesson learned for the future! ;)

To me it seems there are two ways forward:
- either we remove the `success: true/false` paradigm entirely and replace it with status codes. A result with 200 would then be "guaranteed" from the servers point of view to have been treated correctly with no issues,
- or we keep the `success: true/false` flag and only use abort codes for the cases when there is a concrete problem (this would remove most of the 400s ("bad request"), but would keep the 401s ("unauthorized")).

Alternative 1 seems cleaner and probably more in line with common web and REST practices, which makes it attractive, but it is a lot of work both for us on the backend side and for @Endimion75 on the UI side to update the API to this standard, and as @Endimion75 rightly points out, it breaks backwards compatibility.

**Comment by local-minimum** [comment]
@Endimion75 I would say that it isn't clear that the data is correct, if the image isn't griddable by the back-end it isn't correct. The UI must handle this and give the user a way to react. And this is where the sending of grid offsets should help along the request as discussed some months ago would be needed for the request to be processed properly. As I mentioned in #135, this isn't new with @skymandr pull request and to be sure to not miss any discussion I would suggest making sure that you get notifications for all issues and comments on the repository.
You are correct `succes = true` becomes redundant and should probably be deprecated.

**Comment by skymandr** [comment]
Ok, in that case I suggest we prioritise adapting the API backend to use this logic. It's boring but straightforward - perfect pollen season work, so should be doable within days after #135 is landed... ;)

**Comment by local-minimum** [comment]
I think we have discussion of fail-states from the API in 3-5 different issues. If there's need for further discussion, I suggest we take it under a specific issue about how fail-states should look and need to be handled. Otherwise I suggest as @skymandr that whoever rewrites the entire api opens a pullrequest and we discuss the topic on that pull request.

**Comment by skymandr** [comment]
Close?
